### PR TITLE
worker threads, thread pools, and fast map/zip/reduce

### PR DESCRIPTION
**Pull request by MankyDanky**
Key Changes:
- Implement multithreading

### CODE REVIEW

File: packages/tstorch/src/fast_ops.ts [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+        inStrides: Strides,
+    ): void => {
+        const size = shapeProduct(outShape);
+        const aligned =
+            shapesEqual(outShape, inShape) && stridesEqual(outStrides, inStrides);
+
+        const pool = getPool();
+        if (
+            pool &&
+            size >= PARALLEL_THRESHOLD &&
```
Comment (r-chong) [inline packages/tstorch/src/fast_ops.ts:167]: This introduces a second execution path for tensor ops when size >= PARALLEL_THRESHOLD and the storage is a SharedArrayBuffer.

The same operation may now execute either the sequential tensor_ops kernel or the worker-thread kernel, depending on tensor size/storage. In particular, fastTensorReduce will accumulate values in a different floating-point order than the sequential implementation.

The current tests mostly are forward pass correctness. It would also be good to verify training/gradient parity in backprop, which relies on reductions/gradient accumulation (ie., the same model trained right below and right above PARALLEL_THRESHOLD).

Not blocking though, and I can do the test

File: packages/tstorch/src/fast_ops_worker.ts [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+
+const { workerId, syncBuffer } = workerData as {
+    workerId: number;
+    syncBuffer: SharedArrayBuffer;
+};
+const syncArray = new Int32Array(syncBuffer);
+
+function handleMap(task: MapTask): void {
+    const fn = new Function('return ' + task.fnSource)() as (x: number) => number;
```
Comment (r-chong) [inline packages/tstorch/src/fast_ops_worker.ts]: Since the worker reconstructs the operator using fn.toString() and new Function, any captured variables or closures (outer constants or variables) are lost. 

That means tensor ops must depend only on their arguments. Is that an intended constraint? It might be worth documenting, since in the sequential path this isn't a requirement

Comment (MankyDanky) [reply packages/tstorch/src/fast_ops_worker.ts]: Yah this is correct, since all the tensor ops are pure functions this isn't an issue but you are right this probably should be added to a comment

File: packages/tstorch/src/fast_ops.ts (hunk 1 above)
Comment (r-chong) [reply packages/tstorch/src/fast_ops.ts:167]: Ok just verified with some additional tests

```json
{
  weightsBelow: [
    -0.8328319999999957,
    0.16716800000000442,
    1.1671680000000044,
    2.1671680000000046,
    3.1671680000000038,
    4.167168000000004,
    5.167168000000004,
    6.167168000000004
  ],
  weightsAbove: [
    -0.832832000000078,
    0.16716799999992216,
    1.1671679999999223,
    2.167167999999922,
    3.167167999999922,
    4.167167999999922,
    5.167167999999922,
    6.167167999999922
  ]
}
```

Command: `node --no-warnings --loader ./packages/tstorch/scripts/ts-loader.mjs ./packages/demo/index.ts`

Comment (r-chong) [reply packages/tstorch/src/fast_ops.ts:167]: Diff is 8.2x10e-14 (basically equivalent !)

**Review by r-chong — APPROVED** [review]
I just pushed some additional tests, and everything passes - LGTM!
